packages/@uppy/companion-client/src/RequestClient.js

```diff
@@ -94,92 +90,92 @@ export default class RequestClient {
     return `${this.hostname}/${url}`
   }
 
-  #errorHandler (method, path) {
-    return (err) => {
-      if (!err?.isAuthError) {
-        // eslint-disable-next-line no-param-reassign
-        err = new ErrorWithCause(`Could not ${method} ${this.#getUrl(path)}`, { cause: err })
-      }
-      return Promise.reject(err)
+  /*
+    Preflight was added to avoid breaking change between older Companion-client versions and
+    newer Companion versions and vice-versa. Usually the break will manifest via CORS errors because a
+    version of companion-client could be sending certain headers to a version of Companion server that
+    does not support those headers. In which case, the default preflight would lead to CORS.
+    So to avoid those errors, we do preflight ourselves, to see what headers the Companion server
+    we are communicating with allows. And based on that, companion-client knows what headers to
+    send and what headers to not send.
+
+    The preflight only happens once throughout the life-cycle of a certain
+    Companion-client <-> Companion-server pair (allowedHeadersCache).
+    Subsequent requests use the cached result of the preflight.
+    However if there is an error retrieving the allowed headers, we will try again next time
+  */
+  async preflight (path) {
+    const allowedHeadersCached = allowedHeadersCache.get(this.hostname)
+    if (allowedHeadersCached != null) {
+      // The cache may contain a promise that will resolve to undefined if the
+      // fetch call fails. In that case, we want to retry sending a preflight response.
+      const cachedResult = await allowedHeadersCached
+      if (cachedResult != null)  return cachedResult
     }
```

Comment:
```suggestion
    if (allowedHeadersCached != null) return allowedHeadersCached
```

I'm afraid we might be creating a race condition here, by awaiting before running the fetch operation. Can't we instead return fallbackAllowedHeaders instead of undefined inside the catch (after `allowedHeadersCache.delete(this.hostname)`)? Because we still call `allowedHeadersCache.delete` we will re-fetch next time